spark-3.1-spanner-lib/src/main/java/com/google/cloud/spark/spanner/SpannerInputPartitionContext.java

```diff
@@ -52,14 +52,16 @@ public InputPartitionReaderContext<InternalRow> createPartitionReaderContext() {
     if (projectId != null && projectId != "") {
       spannerOptionsBuilder = spannerOptionsBuilder.setProjectId(projectId);
     }
-    SpannerOptions spannerOptions = spannerOptionsBuilder.build();
-    Spanner spanner = spannerOptions.getService();
-    BatchClient batchClient =
-        spanner.getBatchClient(
-            DatabaseId.of(projectId, this.opts.get("instanceId"), this.opts.get("databaseId")));
-    try (BatchReadOnlyTransaction txn =
-        batchClient.batchReadOnlyTransaction(this.batchTransactionId)) {
+    try {
+      SpannerOptions spannerOptions = spannerOptionsBuilder.build();
+      Spanner spanner = spannerOptions.getService();
+      BatchClient batchClient =
+          spanner.getBatchClient(
+              DatabaseId.of(projectId, this.opts.get("instanceId"), this.opts.get("databaseId")));
+      BatchReadOnlyTransaction txn = batchClient.batchReadOnlyTransaction(this.batchTransactionId);
       return new SpannerInputPartitionReaderContext(txn.execute(this.partition));
+    } finally {
+      spanner.close();
```

Comment:
@halio-g this code is invalid because the Spanner spanner object only exists in the scope of the try and not in the finally. I had commented earlier that a while ago I experimented with invoking spanner.close() explicitly in the finally block but sadly by that point, the code using its channels hasn't terminated so the right remedy for us might intead be:

* a runtime finalizer that when it detects its being garbage collected (after the SpannerInputPartitionReaderContext using code) no longer has references to it, explicitly invokes spanner.close() per https://web.mit.edu/java_v1.0.2/www/tutorial/java/javaOO/finalize.html

Reply:
@halio-g Java runtime finalizers were deprecated and instead they recommend that we use AutoCloseable https://docs.oracle.com/javase/10/docs/api/java/lang/AutoCloseable.html in new versions so am gonna prototype and mail a change